### PR TITLE
New Plugin: extkubernetes - Expose External IPs of Load Balancers

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

* Adds a new plugin (_extkubernetes_) that exposes external ip addresses of loadBalancer services
* Starts to moves some code common with _kubernetes_ to `pkg/kubernetes` (e.g. API connection, watch).

"Connection sharing" with the _kubernetes_ plugin is not implemented (that is, if both plugins are active, they do not share the same API connection).  I think the general mode of deployment would be to serve external IPs from a separate instance of CoreDNS, not from the internal cluster DNS, so connection sharing would seldom be a factor.  But if there is a demand to serve the data from the same instance, then we can implement connection sharing later.

No `coredns/ci` test yet.  It will be complicated to implement, but doable.  It will involve setting up virtual routers and load balancers (metalLB) in the framework of the test. I'll start on this only if it looks like the plugin will be accepted.

### 2. Which issues (if any) are related?
#1851
kubernetes/dns#242

### CODE REVIEW

**Comment by chrisohaver** [comment]
FYI, @johnbelamaric

**Comment by chrisohaver** [comment]
I should clarify that the state of this PR is that it is functional. I've tested it using minikube + metalLB, and it works as expected. Hence I am not marking this as a work in progress.
There is still some code duplication, more code could be moved to pkg/kubernetes.

Awaiting feedback and general acceptance of the approach before continuing.

File: plugin.cfg [hunk 1]
```diff
@@ -44,6 +44,7 @@ hosts:hosts
 route53:route53
 federation:federation
 kubernetes:kubernetes
+extkubernetes:extkubernetes
```
Comment (miekg) [inline plugin.cfg:47]: don't really like the name, should we just go for external, or externelkubernetes or ... something?

File: plugin/extkubernetes/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
+# extkubernetes
+
+## Name
+
+*extkubernetes* - enables reading external zone data from a Kubernetes cluster for Load Balancer IPs.
+
+## Description
+
+Creates A/AAAA, SRV, and PTR records for the External IPs of each LoadBalancer type Service in a Kubernetes cluster.
```
Comment (miekg) [inline plugin/extkubernetes/README.md:9]: Please full sentences. The *extkubernernetes* plugin ...

File: plugin/extkubernetes/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    kubeconfig KUBECONFIG CONTEXT
+    namespaces NAMESPACE...
+    labels EXPRESSION
+    ttl TTL
+    transfer to ADDRESS...
+    fallthrough [ZONES...]
+}
+```
+
+* `resyncperiod` specifies the Kubernetes API refresh period to be **DURATION**.
```
Comment (miekg) [inline plugin/extkubernetes/README.md:38]: Just reference the kubernetes plugin here? Removes the need to duplicate all the info here.

File: plugin/extkubernetes/apiproxy.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+type apiProxy struct {
+	http.Server
+	listener net.Listener
+	handler  proxyHandler
+}
+
+func (p *proxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	upstream := p.Select()
+	network := "tcp"
+	address := upstream.Name
```
Comment (miekg) [inline plugin/extkubernetes/apiproxy.go:25]: `address` is only used in the dial just below? No need to put this in a new var

File: plugin/extkubernetes/apiproxy.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+	d, err := net.Dial(network, address)
+	if err != nil {
+		log.Errorf("Unable to establish connection to upstream %s://%s: %s", network, address, err)
+		http.Error(w, fmt.Sprintf("Unable to establish connection to upstream %s://%s: %s", network, address, err), 500)
+		return
+	}
+	hj, ok := w.(http.Hijacker)
```
Comment (miekg) [inline plugin/extkubernetes/apiproxy.go:33]: why is this needed?

File: plugin/extkubernetes/handler.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
+package kubernetes
+
+import (
+	"context"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/request"
```
Comment (miekg) [inline plugin/extkubernetes/handler.go:7]: fmt

File: plugin/extkubernetes/handler.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/miekg/dns"
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	opt := plugin.Options{}
```
Comment (miekg) [inline plugin/extkubernetes/handler.go:14]: where is opt used? Please move this closer

File: plugin/extkubernetes/handler_test.go [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+	"github.com/coredns/coredns/plugin/pkg/kubernetes/object"
+	"github.com/coredns/coredns/plugin/pkg/watch"
+	"github.com/coredns/coredns/plugin/test"
+
+	"github.com/miekg/dns"
+	api "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var dnsTestCases = []test.Case{
```
Comment (miekg) [inline plugin/extkubernetes/handler_test.go:18]: is this just duplicating stuff?

**Comment by miekg** [comment]
This almost duplicates the kubernetes plugin. Please figure out a way to not do that and not fill up pkg/* with kubernetes stuff

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] New Plugin: e..." ]
>Closed #2337.

Note, Chris: I'm very much in favor of having this, but can we do it in much 
less than 3000 lines added?

**Comment by fturib** [comment]
from @johnbelamaric:
> For this reason I think it makes sense to be a different plugin.

from @miekg :
> This almost duplicates the kubernetes plugin. Please figure out a way to not do that and not fill up pkg/* with kubernetes stuff

I guess there is too much constraints here.

If we want to make 2 separates plugins, then we'd better have common code (throuhg lib), and also maybe share the same connection to K8s API.


I am not sure why the number of lines make a difference: I would think the important is to have code that with good design,  and maintainable, and UT (which adds line for sure).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] New Plugin: e..." ]
>I am not sure why the number of lines make a difference: I would think the important is to have code that with good design,  and maintainable, and UT (which adds line for sure).

It's not CLOC, but if duplicated, we need something to keep things in sync 
(automatically)

/Miek

--
Miek Gieben

**Comment by fturib** [comment]
I do not find the meaning for CLOC (https://acronyms.thefreedictionary.com/CLOC). Can you translate ?

> but if duplicated, we need something to keep things in sync

I guess we do not want to duplicate but rather have a common library,  no ?
and the common part, packaged as a lib, is moved to plugin/pkg ... Is it not the purpose of this folder ?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] New Plugin: e..." ]
>I do not find the meaning for CLOC (https://acronyms.thefreedictionary.com/CLOC). Can you translate ?
>
>> but if duplicated, we need something to keep things in sync
>
>I guess we do not want to duplicate but rather have a common library,  no ?
>and the common part, packaged as a lib, is moved to plugin/pkg ... Is it not the purpose of this folder ?

yes, but everything within limits of course; if the entire plugin lives in pkg/ 
you have to ask what you're doing.

Seems doing this in the kubernets plugins seems more sensible?
(don't know what I said before, but 3000 cloc for a little feature seems to be 
pushing it)
